File: frontend/src/lib/api/snmp.ts
import { apiClient } from "./client";

// ============================================================================
// Config types (mirror backend Pydantic models)
// ============================================================================

export interface SNMPListenAddress {
  address: string;
  port: string | null;
}

export interface SNMPCommunity {
  name: string;
  authorization: string | null; // ro | rw
  clients: string[];
  networks: string[];
}

export interface SNMPScriptExtension {
  name: string;
  script: string | null;
}

export interface SNMPTrapTarget {
  address: string;
  community: string | null;
  port: string | null;
}

export interface SNMPv3Auth {
  type: string | null; // md5 | sha
  plaintext_password: string | null;
  encrypted_password: string | null;
}

export interface SNMPv3Privacy {
  type: string | null; // des | aes
  plaintext_password: string | null;
  encrypted_password: string | null;
}

export interface SNMPv3Group {
  name: string;
  mode: string | null; // ro | rw
  seclevel: string | null; // noauth | auth | priv
  view: string | null;
}

export interface SNMPv3User {
  name: string;
  group: string | null;
  mode: string | null; // ro | rw
  auth: SNMPv3Auth | null;
  privacy: SNMPv3Privacy | null;
}

export interface SNMPv3ViewOid {
  oid: string;
  mask: string | null;
  exclude: string[];
}

export interface SNMPv3View {
  name: string;
  oids: SNMPv3ViewOid[];
}

export interface SNMPv3TrapTarget {
  address: string;
  user: string | null;
  type: string | null; // inform | trap
  protocol: string | null; // udp | tcp
  port: string | null;
  auth: SNMPv3Auth | null;
  privacy: SNMPv3Privacy | null;
}

export interface SNMPv3 {
  engineid: string | null;
  groups: SNMPv3Group[];
  users: SNMPv3User[];
  views: SNMPv3View[];
  trap_targets: SNMPv3TrapTarget[];
}

export interface SNMPConfig {
  contact: string | null;
  description: string | null;
  location: string | null;
  protocol: string | null; // udp | tcp
  trap_source: string | null;
  vrf: string | null;
  smux_peers: string[];
  oid_enable: string[];
  listen_addresses: SNMPListenAddress[];
  communities: SNMPCommunity[];
  mib_interfaces: string[];
  mib_interface_max: string | null;
  script_extensions: SNMPScriptExtension[];
  trap_targets: SNMPTrapTarget[];
  v3: SNMPv3;
}

export interface SNMPCapabilities {
  version: string;
  features: {
    snmp: { supported: boolean; description: string };
    protocol: { supported: boolean; description: string; values: string[]; default: string };
    oid_enable: { supported: boolean; description: string; multi_value: boolean; values: string[] };
    smux_peer: { supported: boolean; description: string; multi_value: boolean };
    listen_address: { supported: boolean; description: string; default_port: string };
    community: {
      supported: boolean;
      description: string;
      authorization_values: string[];
      default_authorization: string;
    };
    mib: { supported: boolean; description: string; interface_prefixes: string[] };
    script_extensions: { supported: boolean; description: string };
    trap_target: { supported: boolean; description: string; default_port: string };
    v3: {
      supported: boolean;
      description: string;
      auth_types: string[];
      privacy_types: string[];
      mode_values: string[];
      seclevel_values: string[];
      trap_type_values: string[];
      trap_protocol_values: string[];
      default_trap_port: string;
    };
  };
  version_info: { is_1_4: boolean; is_1_5: boolean };
}

export interface BatchOperation {
  op: string;
  value?: string | null;
}

export interface VyOSResponse {
  success: boolean;
  data?: Record<string, unknown> | null;
  error?: string | null;
}

// ============================================================================
// Update payload types
// ============================================================================

export interface SNMPGeneralUpdate {
  original: SNMPConfig;
  contact: string;
  description: string;
  location: string;
  protocol: string; // "" = default (udp)
  trapSource: string;
  vrf: string;
  engineid: string;
  smuxPeers: string[];
  oidEnable: string[];
  mibInterfaces: string[];
  mibInterfaceMax: string;
}

export interface SNMPListenAddressUpdate {
  address: string;
  port: string; // "" = default (161)
}

export interface SNMPCommunityUpdate {
  name: string;
  authorization: string; // "" = default (ro)
  clients: string[];
  networks: string[];
}

export interface SNMPTrapTargetUpdate {
  address: string;
  community: string;
  port: string; // "" = default (162)
}

export interface SNMPScriptExtensionUpdate {
  name: string;
  script: string; // "" = none
}

export interface SNMPv3CredentialUpdate {
  type: string; // "" = unset
  passwordMode: "plaintext" | "encrypted";
  password: string; // "" = unchanged (edit) / unset (create)
}

export interface SNMPv3UserUpdate {
  name: string;
  group: string;
  mode: string; // "" = default
  authEnabled: boolean;
  auth: SNMPv3CredentialUpdate;
  privacyEnabled: boolean;
  privacy: SNMPv3CredentialUpdate;
}

export interface SNMPv3GroupUpdate {
  name: string;
  mode: string; // "" = default (ro)
  seclevel: string; // "" = default (auth)
  view: string;
}

export interface SNMPv3ViewUpdate {
  name: string;
  oids: SNMPv3ViewOid[];
}

export interface SNMPv3TrapTargetUpdate {
  address: string;
  user: string;
  type: string; // "" = default (inform)
  protocol: string; // "" = default (udp)
  port: string; // "" = default (162)
  authEnabled: boolean;
  auth: SNMPv3CredentialUpdate;
  privacyEnabled: boolean;
  privacy: SNMPv3CredentialUpdate;
}

// ============================================================================
// Helpers
// ============================================================================

/** Build set/delete ops for a single-value field that may be cleared. */
function diffSingle(
  ops: BatchOperation[],
  setOp: string,
  deleteOp: string,
  oldVal: string | null | undefined,
  newVal: string
): void {
  const o = oldVal ?? "";
  if (newVal === o) return;
  if (newVal === "") ops.push({ op: deleteOp });
  else ops.push({ op: setOp, value: newVal });
}

/** Build set/delete ops for a single-value field scoped to one entity (e.g. "addr"). */
function diffSingleScoped(
  ops: BatchOperation[],
  setOp: string,
  deleteOp: string,
  scope: string,
  oldVal: string | null | undefined,
  newVal: string
): void {
  const o = oldVal ?? "";
  if (newVal === o) return;
  if (newVal === "") ops.push({ op: deleteOp, value: scope });
  else ops.push({ op: setOp, value: `${scope},${newVal}` });
}

/** Build set/delete ops for a scoped multi-value list (e.g. community clients). */
function diffMultiScoped(
  ops: BatchOperation[],
  setOp: string,
  deleteOp: string,
  scope: string,
  oldVals: string[],
  newVals: string[]
): void {
  for (const v of oldVals.filter((x) => !newVals.includes(x))) {
    ops.push({ op: deleteOp, value: `${scope},${v}` });
  }
  for (const v of newVals.filter((x) => !oldVals.includes(x))) {
    ops.push({ op: setOp, value: `${scope},${v}` });
  }
}

/** Build set/delete ops for a global multi-value list. */
function diffMulti(
  ops: BatchOperation[],
  setOp: string,
  deleteOp: string,
  oldVals: string[],
  newVals: string[]
): void {
  for (const v of oldVals.filter((x) => !newVals.includes(x))) {
    ops.push({ op: deleteOp, value: v });
  }
  for (const v of newVals.filter((x) => !oldVals.includes(x))) {
    ops.push({ op: setOp, value: v });
  }
}

// ============================================================================
// Service
// ============================================================================

class SNMPService {
  async getCapabilities(): Promise<SNMPCapabilities> {
    return apiClient.get<SNMPCapabilities>("/vyos/snmp/capabilities");
  }

  async getConfig(refresh = false): Promise<SNMPConfig> {
    return apiClient.get<SNMPConfig>("/vyos/snmp/config", {
      refresh: refresh.toString(),
    });
  }

  private async batch(operations: BatchOperation[]): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/snmp/batch", {
      operations,
    });
    if (!result.success) throw new Error(result.error || "Operation failed");
    return result;
  }

  // -------------------------------------------------------------- General
  async updateGeneral(u: SNMPGeneralUpdate): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    const o = u.original;

    diffSingle(ops, "set_contact", "delete_contact", o.contact, u.contact.trim());
    diffSingle(ops, "set_description", "delete_description", o.description, u.description.trim());
    diffSingle(ops, "set_location", "delete_location", o.location, u.location.trim());
    diffSingle(ops, "set_protocol", "delete_protocol", o.protocol, u.protocol);
    diffSingle(ops, "set_trap_source", "delete_trap_source", o.trap_source, u.trapSource.trim());
    diffSingle(ops, "set_vrf", "delete_vrf", o.vrf, u.vrf.trim());
    diffSingle(ops, "set_v3_engineid", "delete_v3_engineid", o.v3.engineid, u.engineid.trim());
    diffSingle(
      ops,
      "set_mib_interface_max",
      "delete_mib_interface_max",
      o.mib_interface_max,
      u.mibInterfaceMax.trim()
    );

    diffMulti(ops, "set_smux_peer", "delete_smux_peer", o.smux_peers, u.smuxPeers);
    diffMulti(ops, "set_oid_enable", "delete_oid_enable", o.oid_enable, u.oidEnable);
    diffMulti(ops, "set_mib_interface", "delete_mib_interface", o.mib_interfaces, u.mibInterfaces);

    if (ops.length === 0) return { success: true };
    return this.batch(ops);
  }

  // ------------------------------------------------------- Listen address
  async saveListenAddress(
    original: SNMPListenAddress | null,
    u: SNMPListenAddressUpdate
  ): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    const addr = u.address.trim();
    if (original === null) ops.push({ op: "set_listen_address", value: addr });
    diffSingleScoped(
      ops,
      "set_listen_address_port",
      "delete_listen_address_port",
      addr,
      original?.port,
      u.port.trim()
    );
    if (ops.length === 0) return { success: true };
    return this.batch(ops);
  }

  async deleteListenAddress(address: string): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_listen_address", value: address }]);
  }

  // ----------------------------------------------------------- Community
  async saveCommunity(
    original: SNMPCommunity | null,
    u: SNMPCommunityUpdate
  ): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    const name = u.name.trim();
    if (original === null) ops.push({ op: "set_community", value: name });

    diffSingleScoped(
      ops,
      "set_community_authorization",
      "delete_community_authorization",
      name,
      original?.authorization,
      u.authorization
    );
    diffMultiScoped(
      ops,
      "set_community_client",
      "delete_community_client",
      name,
      original?.clients ?? [],
      u.clients
    );
    diffMultiScoped(
      ops,
      "set_community_network",
      "delete_community_network",
      name,
      original?.networks ?? [],
      u.networks
    );

    if (ops.length === 0) return { success: true };
    return this.batch(ops);
  }

  async deleteCommunity(name: string): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_community", value: name }]);
  }

  // -------------------------------------------- Trap target (v1/v2c)
  async saveTrapTarget(
    original: SNMPTrapTarget | null,
    u: SNMPTrapTargetUpdate
  ): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    const addr = u.address.trim();
    if (original === null) ops.push({ op: "set_trap_target", value: addr });

    diffSingleScoped(
      ops,
      "set_trap_target_community",
      "delete_trap_target_community",
      addr,
      original?.community,
      u.community.trim()
    );
    diffSingleScoped(
      ops,
      "set_trap_target_port",
      "delete_trap_target_port",
      addr,
      original?.port,
      u.port.trim()
    );

    if (ops.length === 0) return { success: true };
    return this.batch(ops);
  }

  async deleteTrapTarget(address: string): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_trap_target", value: address }]);
  }

  // -------------------------------------------------- Script extension
  async saveScriptExtension(
    original: SNMPScriptExtension | null,
    u: SNMPScriptExtensionUpdate
  ): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    const name = u.name.trim();
    if (original === null) ops.push({ op: "set_script_extension", value: name });
    diffSingleScoped(
      ops,
      "set_script_extension_script",
      "delete_script_extension_script",
      name,
      original?.script,
      u.script.trim()
    );
    if (ops.length === 0) return { success: true };
    return this.batch(ops);
  }

  async deleteScriptExtension(name: string): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_script_extension", value: name }]);
  }

  // ----------------------------------------------------------- v3 user
  async saveV3User(
    original: SNMPv3User | null,
    u: SNMPv3UserUpdate
  ): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    const name = u.name.trim();
    if (original === null) ops.push({ op: "set_v3_user", value: name });

    diffSingleScoped(ops, "set_v3_user_group", "delete_v3_user_group", name, original?.group, u.group.trim());
    diffSingleScoped(ops, "set_v3_user_mode", "delete_v3_user_mode", name, original?.mode, u.mode);

    this.diffCredential(
      ops,
      name,
      "v3_user_auth",
      original?.auth ?? null,
      u.authEnabled ? u.auth : null
    );
    this.diffCredential(
      ops,
      name,
      "v3_user_privacy",
      original?.privacy ?? null,
      u.privacyEnabled ? u.privacy : null
    );

    if (ops.length === 0) return { success: true };
    return this.batch(ops);
  }

  async deleteV3User(name: string): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_v3_user", value: name }]);
  }

  // ---------------------------------------------------------- v3 group
  async saveV3Group(
    original: SNMPv3Group | null,
    u: SNMPv3GroupUpdate
  ): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    const name = u.name.trim();
    if (original === null) ops.push({ op: "set_v3_group", value: name });

    diffSingleScoped(ops, "set_v3_group_mode", "delete_v3_group_mode", name, original?.mode, u.mode);
    diffSingleScoped(ops, "set_v3_group_seclevel", "delete_v3_group_seclevel", name, original?.seclevel, u.seclevel);
    diffSingleScoped(ops, "set_v3_group_view", "delete_v3_group_view", name, original?.view, u.view.trim());

    if (ops.length === 0) return { success: true };
    return this.batch(ops);
  }

  async deleteV3Group(name: string): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_v3_group", value: name }]);
  }

  // ----------------------------------------------------------- v3 view
  async saveV3View(
    original: SNMPv3View | null,
    u: SNMPv3ViewUpdate
  ): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    const name = u.name.trim();
    if (original === null) ops.push({ op: "set_v3_view", value: name });

    const origOids = original?.oids ?? [];
    const newOidKeys = u.oids.map((o) => o.oid);

    // Removed OIDs
    for (const o of origOids.filter((x) => !newOidKeys.includes(x.oid))) {
      ops.push({ op: "delete_v3_view_oid", value: `${name},${o.oid}` });
    }
    // Added / changed OIDs
    for (const oid of u.oids) {
      const prev = origOids.find((x) => x.oid === oid.oid) ?? null;
      if (prev === null) ops.push({ op: "set_v3_view_oid", value: `${name},${oid.oid}` });
      // mask
      diffSingleScoped(
        ops,
        "set_v3_view_oid_mask",
        "delete_v3_view_oid_mask",
        `${name},${oid.oid}`,
        prev?.mask,
        (oid.mask ?? "").trim()
      );
      // excludes (scoped to view,oid)
      diffMultiScoped(
        ops,
        "set_v3_view_oid_exclude",
        "delete_v3_view_oid_exclude",
        `${name},${oid.oid}`,
        prev?.exclude ?? [],
        oid.exclude
      );
    }

    if (ops.length === 0) return { success: true };
    return this.batch(ops);
  }

  async deleteV3View(name: string): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_v3_view", value: name }]);
  }

  // -------------------------------------------------- v3 trap target
  async saveV3TrapTarget(
    original: SNMPv3TrapTarget | null,
    u: SNMPv3TrapTargetUpdate
  ): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    const addr = u.address.trim();
    if (original === null) ops.push({ op: "set_v3_trap_target", value: addr });

    diffSingleScoped(ops, "set_v3_trap_target_user", "delete_v3_trap_target_user", addr, original?.user, u.user.trim());
    diffSingleScoped(ops, "set_v3_trap_target_type", "delete_v3_trap_target_type", addr, original?.type, u.type);
    diffSingleScoped(ops, "set_v3_trap_target_protocol", "delete_v3_trap_target_protocol", addr, original?.protocol, u.protocol);
    diffSingleScoped(ops, "set_v3_trap_target_port", "delete_v3_trap_target_port", addr, original?.port, u.port.trim());

    this.diffCredential(
      ops,
      addr,
      "v3_trap_target_auth",
      original?.auth ?? null,
      u.authEnabled ? u.auth : null
    );
    this.diffCredential(
      ops,
      addr,
      "v3_trap_target_privacy",
      original?.privacy ?? null,
      u.privacyEnabled ? u.privacy : null
    );

    if (ops.length === 0) return { success: true };
    return this.batch(ops);
  }

  async deleteV3TrapTarget(address: string): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_v3_trap_target", value: address }]);
  }

  // ----------------------------------------------------------- internal
  /**
   * Diff an auth/privacy credential block. `prefix` is the builder method stem,
   * e.g. "v3_user_auth" -> set_v3_user_auth_type / *_plaintext_password / ...
   * Passwords are only written when a (non-empty) new value is supplied, since
   * VyOS hashes plaintext on commit and never returns it for comparison.
   */
  private diffCredential(
    ops: BatchOperation[],
    scope: string,
    prefix: string,
    original: SNMPv3Auth | SNMPv3Privacy | null,
    next: SNMPv3CredentialUpdate | null
  ): void {
    // Disabling the credential block entirely.
    if (next === null) {
      if (original?.type) ops.push({ op: `delete_${prefix}_type`, value: scope });
      if (original?.plaintext_password)
        ops.push({ op: `delete_${prefix}_plaintext_password`, value: scope });
      if (original?.encrypted_password)
        ops.push({ op: `delete_${prefix}_encrypted_password`, value: scope });
      return;
    }

    diffSingleScoped(
      ops,
      `set_${prefix}_type`,
      `delete_${prefix}_type`,
      scope,
      original?.type,
      next.type
    );

    const pw = next.password.trim();
    if (pw !== "") {
      if (next.passwordMode === "plaintext") {
        ops.push({ op: `set_${prefix}_plaintext_password`, value: `${scope},${pw}` });
        if (original?.encrypted_password)
          ops.push({ op: `delete_${prefix}_encrypted_password`, value: scope });
      } else {
        ops.push({ op: `set_${prefix}_encrypted_password`, value: `${scope},${pw}` });
        if (original?.plaintext_password)
          ops.push({ op: `delete_${prefix}_plaintext_password`, value: scope });
      }
    }
  }
}

export const snmpService = new SNMPService();

File: backend/routers/snmp/snmp.py
"""SNMP Service Router.

API endpoints for managing VyOS SNMP (Simple Network Management Protocol)
configuration.

Version differences:
  1.4 and 1.5 — identical SNMP configuration paths; no version-specific behavior.

Endpoints:
  GET  /vyos/snmp/capabilities  — version-aware feature flags
  GET  /vyos/snmp/config        — normalized SNMP configuration
  POST /vyos/snmp/batch         — atomic set/delete operations
"""

from fastapi import APIRouter, HTTPException, Request
from starlette.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from session_vyos_service import get_session_vyos_service
from vyos_builders.snmp import SNMPBatchBuilder
from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
import inspect
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/vyos/snmp", tags=["snmp"])


# ============================================================================
# Pydantic Models
# ============================================================================


class SNMPListenAddress(BaseModel):
    """An address SNMP listens on, with an optional custom port."""
    address: str
    port: Optional[str] = None


class SNMPCommunity(BaseModel):
    """An SNMPv1/v2c community."""
    name: str
    authorization: Optional[str] = None  # ro|rw (default ro)
    clients: List[str] = []
    networks: List[str] = []


class SNMPScriptExtension(BaseModel):
    """A custom script extension for the SNMP agent."""
    name: str
    script: Optional[str] = None


class SNMPTrapTarget(BaseModel):
    """An SNMPv1/v2c trap target."""
    address: str
    community: Optional[str] = None
    port: Optional[str] = None


class SNMPv3Auth(BaseModel):
    """SNMPv3 authentication settings."""
    type: Optional[str] = None  # md5|sha
    plaintext_password: Optional[str] = None
    encrypted_password: Optional[str] = None


class SNMPv3Privacy(BaseModel):
    """SNMPv3 privacy (encryption) settings."""
    type: Optional[str] = None  # des|aes
    plaintext_password: Optional[str] = None
    encrypted_password: Optional[str] = None


class SNMPv3Group(BaseModel):
    """An SNMPv3 group."""
    name: str
    mode: Optional[str] = None  # ro|rw
    seclevel: Optional[str] = None  # noauth|auth|priv
    view: Optional[str] = None


class SNMPv3User(BaseModel):
    """An SNMPv3 user."""
    name: str
    group: Optional[str] = None
    mode: Optional[str] = None  # ro|rw
    auth: Optional[SNMPv3Auth] = None
    privacy: Optional[SNMPv3Privacy] = None


class SNMPv3ViewOid(BaseModel):
    """A single OID subtree within an SNMPv3 view."""
    oid: str
    mask: Optional[str] = None
    exclude: List[str] = []


class SNMPv3View(BaseModel):
    """An SNMPv3 view (collection of OID subtrees)."""
    name: str
    oids: List[SNMPv3ViewOid] = []


class SNMPv3TrapTarget(BaseModel):
    """An SNMPv3 trap/inform target."""
    address: str
    user: Optional[str] = None
    type: Optional[str] = None  # inform|trap
    protocol: Optional[str] = None  # udp|tcp
    port: Optional[str] = None
    auth: Optional[SNMPv3Auth] = None
    privacy: Optional[SNMPv3Privacy] = None


class SNMPv3(BaseModel):
    """SNMPv3 configuration."""
    engineid: Optional[str] = None
    groups: List[SNMPv3Group] = []
    users: List[SNMPv3User] = []
    views: List[SNMPv3View] = []
    trap_targets: List[SNMPv3TrapTarget] = []


class SNMPConfig(BaseModel):
    """Full SNMP service configuration."""
    contact: Optional[str] = None
    description: Optional[str] = None
    location: Optional[str] = None
    protocol: Optional[str] = None  # udp|tcp
    trap_source: Optional[str] = None
    vrf: Optional[str] = None
    smux_peers: List[str] = []
    oid_enable: List[str] = []
    listen_addresses: List[SNMPListenAddress] = []
    communities: List[SNMPCommunity] = []
    mib_interfaces: List[str] = []
    mib_interface_max: Optional[str] = None
    script_extensions: List[SNMPScriptExtension] = []
    trap_targets: List[SNMPTrapTarget] = []
    v3: SNMPv3 = SNMPv3()


class SNMPBatchOperation(BaseModel):
    op: str = Field(..., description="Builder method name")
    value: Optional[str] = Field(
        None,
        description=(
            "Argument(s) for the operation. "
            "Single-arg methods: plain string. "
            "Multi-arg methods: comma-separated (e.g., 'community,ro')."
        ),
    )


class SNMPBatchRequest(BaseModel):
    operations: List[SNMPBatchOperation]


class VyOSResponse(BaseModel):
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# ============================================================================
# Internal builder method denylist
# ============================================================================

_INTERNAL_BUILDER_METHODS = frozenset({
    "add_set", "add_delete", "get_operations", "is_empty",
    "get_capabilities", "mappers", "version", "_operations", "m",
})


# ============================================================================
# Endpoint 1: Capabilities
# ============================================================================


@router.get("/capabilities")
async def get_snmp_capabilities(request: Request):
    """Return SNMP feature capabilities based on the VyOS version."""
    await require_read_permission(request, FeatureGroup.SNMP)
    try:
        service = get_session_vyos_service(request)
        builder = SNMPBatchBuilder(version=service.get_version())
        caps = builder.get_capabilities()
        if hasattr(request.state, "instance") and request.state.instance:
            caps["instance_name"] = request.state.instance.get("name")
            caps["instance_id"] = request.state.instance.get("id")
        return caps
    except Exception:
        logger.exception("Unhandled error in get_snmp_capabilities")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Endpoint 2: Config
# ============================================================================


@router.get("/config", response_model=SNMPConfig)
async def get_snmp_config(http_request: Request, refresh: bool = False):
    """Return the full SNMP configuration in a normalized format."""
    await require_read_permission(http_request, FeatureGroup.SNMP)
    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh=refresh)

        snmp_raw = full_config.get("service", {}).get("snmp", {})
        if not snmp_raw or not isinstance(snmp_raw, dict):
            return SNMPConfig()

        return SNMPConfig(
            contact=snmp_raw.get("contact"),
            description=snmp_raw.get("description"),
            location=snmp_raw.get("location"),
            protocol=snmp_raw.get("protocol"),
            trap_source=snmp_raw.get("trap-source"),
            vrf=snmp_raw.get("vrf"),
            smux_peers=_parse_multi_value(snmp_raw.get("smux-peer")),
            oid_enable=_parse_multi_value(snmp_raw.get("oid-enable")),
            listen_addresses=_parse_listen_addresses(snmp_raw),
            communities=_parse_communities(snmp_raw),
            mib_interfaces=_parse_multi_value(snmp_raw.get("mib", {}).get("interface")),
            mib_interface_max=snmp_raw.get("mib", {}).get("interface-max"),
            script_extensions=_parse_script_extensions(snmp_raw),
            trap_targets=_parse_trap_targets(snmp_raw),
            v3=_parse_v3(snmp_raw.get("v3", {})),
        )
    except Exception:
        logger.exception("Unhandled error in get_snmp_config")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Endpoint 3: Batch Operations
# ============================================================================


@router.post("/batch", response_model=VyOSResponse)
async def snmp_batch_configure(http_request: Request, body: SNMPBatchRequest):
    """Execute a batch of SNMP configuration operations atomically."""
    await require_write_permission(http_request, FeatureGroup.SNMP)
    try:
        service = get_session_vyos_service(http_request)
        builder = SNMPBatchBuilder(version=service.get_version())

        for operation in body.operations:
            if operation.op in _INTERNAL_BUILDER_METHODS or operation.op.startswith("_"):
                raise HTTPException(
                    status_code=400,
                    detail=f"Operation not allowed: {operation.op}",
                )

            method = getattr(builder, operation.op)
            sig = inspect.signature(method)
            params = [p for p in sig.parameters.keys() if p != "self"]

            if len(params) == 0:
                method()
            elif len(params) == 1:
                if operation.value is not None:
                    method(operation.value)
                else:
                    method()
            elif len(params) >= 2:
                if operation.value and "," in operation.value:
                    parts = operation.value.split(",", len(params) - 1)
                    method(*parts)
                elif operation.value:
                    method(operation.value)

        response = service.execute_batch(builder)
        return VyOSResponse(
            success=response.status == 200,
            data={"message": "SNMP configuration updated"},
            error=response.error if response.error else None,
        )
    except HTTPException:
        raise
    except AttributeError as e:
        raise HTTPException(status_code=400, detail=f"Unknown operation: {e}")
    except Exception:
        logger.exception("Unhandled error in snmp_batch_configure")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Config parsers
# ============================================================================


def _parse_multi_value(raw) -> List[str]:
    """Parse a VyOS multi-value node into a sorted list of strings."""
    if raw is None:
        return []
    if isinstance(raw, list):
        return sorted(str(v) for v in raw)
    if isinstance(raw, dict):
        return sorted(raw.keys())
    return [str(raw)]


def _parse_listen_addresses(snmp_raw: dict) -> List[SNMPListenAddress]:
    listen_raw = snmp_raw.get("listen-address", {})
    if not isinstance(listen_raw, dict):
        return []
    result = []
    for address, cfg in listen_raw.items():
        cfg = cfg or {}
        result.append(SNMPListenAddress(address=address, port=cfg.get("port")))
    result.sort(key=lambda a: a.address)
    return result


def _parse_communities(snmp_raw: dict) -> List[SNMPCommunity]:
    comm_raw = snmp_raw.get("community", {})
    if not isinstance(comm_raw, dict):
        return []
    result = []
    for name, cfg in comm_raw.items():
        cfg = cfg or {}
        result.append(SNMPCommunity(
            name=name,
            authorization=cfg.get("authorization"),
            clients=_parse_multi_value(cfg.get("client")),
            networks=_parse_multi_value(cfg.get("network")),
        ))
    result.sort(key=lambda c: c.name)
    return result


def _parse_script_extensions(snmp_raw: dict) -> List[SNMPScriptExtension]:
    ext_raw = snmp_raw.get("script-extensions", {}).get("extension-name", {})
    if not isinstance(ext_raw, dict):
        return []
    result = []
    for name, cfg in ext_raw.items():
        cfg = cfg or {}
        result.append(SNMPScriptExtension(name=name, script=cfg.get("script")))
    result.sort(key=lambda e: e.name)
    return result


def _parse_trap_targets(snmp_raw: dict) -> List[SNMPTrapTarget]:
    trap_raw = snmp_raw.get("trap-target", {})
    if not isinstance(trap_raw, dict):
        return []
    result = []
    for address, cfg in trap_raw.items():
        cfg = cfg or {}
        result.append(SNMPTrapTarget(
            address=address,
            community=cfg.get("community"),
            port=cfg.get("port"),
        ))
    result.sort(key=lambda t: t.address)
    return result


def _parse_auth(cfg: dict) -> Optional[SNMPv3Auth]:
    auth_raw = cfg.get("auth")
    if not isinstance(auth_raw, dict):
        return None
    return SNMPv3Auth(
        type=auth_raw.get("type"),
        plaintext_password=auth_raw.get("plaintext-password"),
        encrypted_password=auth_raw.get("encrypted-password"),
    )


def _parse_privacy(cfg: dict) -> Optional[SNMPv3Privacy]:
    priv_raw = cfg.get("privacy")
    if not isinstance(priv_raw, dict):
        return None
    return SNMPv3Privacy(
        type=priv_raw.get("type"),
        plaintext_password=priv_raw.get("plaintext-password"),
        encrypted_password=priv_raw.get("encrypted-password"),
    )


def _parse_v3(v3_raw: dict) -> SNMPv3:
    if not isinstance(v3_raw, dict) or not v3_raw:
        return SNMPv3()

    groups = []
    for name, cfg in (v3_raw.get("group", {}) or {}).items():
        cfg = cfg or {}
        groups.append(SNMPv3Group(
            name=name,
            mode=cfg.get("mode"),
            seclevel=cfg.get("seclevel"),
            view=cfg.get("view"),
        ))
    groups.sort(key=lambda g: g.name)

    users = []
    for name, cfg in (v3_raw.get("user", {}) or {}).items():
        cfg = cfg or {}
        users.append(SNMPv3User(
            name=name,
            group=cfg.get("group"),
            mode=cfg.get("mode"),
            auth=_parse_auth(cfg),
            privacy=_parse_privacy(cfg),
        ))
    users.sort(key=lambda u: u.name)

    views = []
    for name, cfg in (v3_raw.get("view", {}) or {}).items():
        cfg = cfg or {}
        oids = []
        for oid, oid_cfg in (cfg.get("oid", {}) or {}).items():
            oid_cfg = oid_cfg or {}
            oids.append(SNMPv3ViewOid(
                oid=oid,
                mask=oid_cfg.get("mask"),
                exclude=_parse_multi_value(oid_cfg.get("exclude")),
            ))
        oids.sort(key=lambda o: o.oid)
        views.append(SNMPv3View(name=name, oids=oids))
    views.sort(key=lambda v: v.name)

    trap_targets = []
    for address, cfg in (v3_raw.get("trap-target", {}) or {}).items():
        cfg = cfg or {}
        trap_targets.append(SNMPv3TrapTarget(
            address=address,
            user=cfg.get("user"),
            type=cfg.get("type"),
            protocol=cfg.get("protocol"),
            port=cfg.get("port"),
            auth=_parse_auth(cfg),
            privacy=_parse_privacy(cfg),
        ))
    trap_targets.sort(key=lambda t: t.address)

    return SNMPv3(
        engineid=v3_raw.get("engineid"),
        groups=groups,
        users=users,
        views=views,
        trap_targets=trap_targets,
    )
